File: 02_selection_sort/selection_sort.js
'use strict';

function findSmallestIndex(array) {
    let smallestElement = array[0];
    let smallestIndex = 0;

    for (let i = 1; i < array.length; i++) {
        if (array[i] < smallestElement) {
            smallestElement = array[i];
            smallestIndex = i;
        }
    }

    return smallestIndex;
}

function selectionSort(array) {
    let sortedArray = [];
    let length = array.length;

    for (let i = 0; i < length; i++) {
        let smallestIndex = findSmallestIndex(array);
        sortedArray.push(array.splice(smallestIndex, 1)[0]);
    }

    return sortedArray;
}

console.log(selectionSort([5, 3, 6, 2, 9])); // [2, 3, 5, 6, 9]

File: 02_selection_sort/selection_sort.go
package main

import "fmt"

func findSmallestIndex(arr []int) int {
    smallestElement := arr[0]
    smallestIndex := 0

    for i := 0; i < len(arr); i++ {
        if arr[i] < smallestElement {
            smallestElement = arr[i]
            smallestIndex = i
        }
    }
    return smallestIndex
}

func selectionSort(arr []int) []int {
    length := len(arr)
    sortedArray := make([]int, length)

    for i := 0; i < length; i++ {
        smallestIndex := findSmallestIndex(arr)
        sortedArray[i] = arr[smallestIndex]
        arr = append(arr[:smallestIndex], arr[smallestIndex+1:]...)
    }
    return sortedArray
}

func main() {
    array := []int{5, 3, 6, 2, 9}
    fmt.Println(selectionSort(array))
}
